src/Bicep.Core/Configuration/ConfigurationManager.cs

```diff
@@ -0,0 +1,99 @@
+
+        private static readonly Lazy<RootConfiguration> builtInConfigurationWithAnalyzersDisabledLazy =
+            new(() => RootConfiguration.Bind(builtInRawConfiguration, BuiltInConfigurationResourceName, disableAnalyzers: true));
+
+        private readonly IFileSystem fileSystem;
+
+        public ConfigurationManager(IFileSystem fileSystem)
+        {
+            this.fileSystem = fileSystem;
+        }
+
+        public RootConfiguration GetBuiltInConfiguration(bool disableAnalyzers = false) => disableAnalyzers
+            ? builtInConfigurationWithAnalyzersDisabledLazy.Value
+            : builtInConfigurationLazy.Value;
+
+        public RootConfiguration GetConfiguration(Uri sourceFileUri)
+        {
+            var customConfigurationPath = DiscoverConfigurationFile(fileSystem.Path.GetDirectoryName(sourceFileUri.LocalPath));
+
+            if (customConfigurationPath is not null)
+            {
+                IConfiguration? customConfiguration;
+
+                try
+                {
+                    using var stream = fileSystem.FileStream.Create(customConfigurationPath, FileMode.Open, FileAccess.Read);
+                    var builder = new ConfigurationBuilder()
+                        .AddInMemoryCollection(builtInRawConfiguration.AsEnumerable())
+                        .AddJsonStream(stream);
+
+                    customConfiguration = builder.Build();
+                }
+                catch (Exception exception)
+                {
+                    throw new ConfigurationException(customConfigurationPath, exception.Message);
+                }
+
+                return RootConfiguration.Bind(customConfiguration, customConfigurationPath);
+            }
+
+            return GetBuiltInConfiguration();
+        }
+
+        private static IConfiguration BuildBuiltInRawConfiguration()
+        {
+            using var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream(BuiltInConfigurationResourceName);
+            Debug.Assert(stream is not null, "Default configuration file should exist as embedded resource.");
+
+            var builder = new ConfigurationBuilder();
+            builder.AddJsonStream(stream);
+
+            return builder.Build();
+        }
+
+        private string? DiscoverConfigurationFile(string? currentDirectory)
+        {
+            try
+            {
+                while (!string.IsNullOrEmpty(currentDirectory))
+                {
+                    var configurationPath = this.fileSystem.Path.Combine(currentDirectory, LanguageConstants.BicepConfigurationFileName);
+
+                    if (this.fileSystem.File.Exists(configurationPath))
+                    {
+                        return configurationPath;
+                    }
+
+                    currentDirectory = this.fileSystem.Directory.GetParent(currentDirectory)?.FullName;
+                }
+            }
+            catch (Exception)
+            {
+                return null;
```

Comment:
This feels a bit problematic, although surfacing the exception to the user may also end up showing false positives (e.g. if the user account has missing permissions on a parent dir). I think it might be worth creating an issue to try and get telemetry on this. As a user I feel like I'd rather understand if there was an issue loading my config - without this the experience may appear unpredictable.

Reply:
This was the original code in `ConfigHelper`, but I think you are right that we should definitely not drop the exception silently and return `null`.  I've updated the logic so that any IO related exception will converted to a `ConfigurationException` with an error message like "Error while discovering Bicep configuration file: \"{exception.message}\".".